### PR TITLE
Add iwear_logger wrapper device and AddInstallRPathSupport cmake module

### LINKED ISSUE #112: Add possibility to log wearable data to file
**Is your feature request related to a problem? Please describe.**

Currently we are using the wearable implementation in the [robotology/human-dynamics-estimation](https://github.com/robotology/human-dynamics-estimation) pipeline to access the human-worn sensors data like IMU, FT shoes measurements, etc by providing relevant implementations for the `iWear` interface. The desired problem in hand is to be able to log these sensor data available through the `iWear` interface into a file (mat file), so that it could improve debugging/testing through offline processing. 

**Describe the solution you'd like**
The [`iWearWrapper`](https://github.com/robotology/wearables/blob/master/wrappers/IWear/include/IWearWrapper.h) provides a possibility to pack multiple sensor measurements and pass it through a port. Apart from passing this wrapped data through a port using the thrift data structure, it would also be beneficial if we can optionally log these data to a file at the [`iWearWrapper`](https://github.com/robotology/wearables/blob/master/wrappers/IWear/include/IWearWrapper.h) level. We could possibly add a flag that will allow us to optionally log this data. 

An useful dependency for this purpose could be [yarp-telemetry](https://github.com/robotology/yarp-telemetry) which allows periodic saving and also uses a memory-efficient circular buffer based implementation. However, the limitation is that we can only save the data as channels of vectors (cannot store data as a struct), but this is only a semantic problem.

**Describe alternatives you've considered**
I had tried to add a [dic-iit/matio-cpp](https://github.com/dic-iit/matio-cpp)-based logging at a consumer level (a client device that attaches to a iWearRemapper, for example [HumanStateProvider](https://github.com/robotology/human-dynamics-estimation/tree/master/devices/HumanStateProvider) in [robotology/human-dynamics-estimation](https://github.com/robotology/human-dynamics-estimation)). However functional, this implementation has the limitations of not having a periodic save and also uses a dynamic memory allocation by resizing the vectors all the time. So this might not be suitable for a long-term logging. Further, there is also the delay introduced from passing the information from the producer to the consumer.

It would be helpful to have this logging at a higher level as a generic implementation, which can be extended for any suitable implementations using the `iWearWrapper` interface.

I had a discussion with @Yeshasvitvs regarding the logging and we came to an understanding that `iWearWrapper` level logging could be a possibly feasible solution.
cc  @lrapetti  @traversaro @S-Dafarra @diegoferigo Let me know about your opinion.

Also please correct me if my understanding of the underlying software architecture seems incorrect.

### PR DESCRIPTION
**Pull request by prashanthr05**
This PR,
- adds an `AddInstallRPATHSupport` cmake module that was required to load the shared objects privately linked by the YARP device plugin.
- adds an `iwear_logger` wrapper device that can be attached to any wearable producer to log the data from all the sensors associated with the producer. 

At the heart of its implementation, It replaces,
> the `iwear_wrapper`'s logic of encapsulating the wearable sensors data into a WearableData msg and passing through a YARP port 

with 

> `yarp-telemetry`'s logic of dumping the data into a mat file.

A template for configuring the `iwear_logger` wrapper device can be as follows,

``` xml
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE robot PUBLIC "-//YARP//DTD yarprobotinterface 3.0//EN" "http://www.yarp.it/DTD/yarprobotinterfaceV3.0.dtd">
<robot name="robot-name" build=0 portprefix="">
    <!--Producer Device -->
    <device type="producer_name" name="ProducerDevice">
        <!-- producer parameters go here -->
    </device>
    
    <device type="iwear_logger" name="ProducerLoggerDevice">
        <param name="period">0.01</param>
        
        <param name="logAllQuantities">false</param>
        <param name="logAccelerometers">false</param>
        <param name="logEMGSensors">false</param>
        <param name="logForce3DSensors">false</param>
        <param name="logForceTorque6DSensors">true</param>
        <param name="logFreeBodyAccelerationSensors">true</param>
        <param name="logGyroscopes">true</param>
        <param name="logMagnetometers">false</param>
        <param name="logOrientationSensors">true</param>
        <param name="logPoseSensors">false</param>
        <param name="logPositionSensors">false</param>
        <param name="logTemperatureSensors">false</param>
        <param name="logTorque3DSensors">false</param>
        <param name="logVirtualLinkKinSensors">false</param>
        <param name="logVirtualJointKinSensors">false</param>
        <param name="logVirtualSphericalJointKinSensors">false</param>                
        
        <param name="saveBufferManagerConfiguration">true</param>
        <param name="experimentName">test_iwear_logger</param>
        <param name="path">/path/to/some/folder/</param>
        <param name="n_samples">100000</param>
        <param name="save_periodically">true</param>
        <param name="save_period">120.0</param>
        <param name="data_threshold">300</param>
        <param name="auto_save">true</param>

        <action phase="startup" level="5" type="attach">
            <paramlist name="networks">
            	<!-- attach the source device here -->
                <elem name="ProducerDeviceWrapperLabel"> ProducerDevice </elem>
            </paramlist>
        </action>
        <action phase="shutdown" level="5" type="detach"/>
    </device>

</robot>
```

The data is saved as a  `<experimentName>_<timestamp>.mat` in the specified `<path>` parameter. 
This mat file contains a struct called `<experimentName>`

The sensors data are stored as a struct one for each wearable sensor. The name of this struct is a modified format of Wearable sensor name accounting for Matlab variable name handling. (Special characters like `@/()#` are converted to underscores and the wearable separator is converted to underscore.)
```
<producer-device-name>_<sensor-type>_<sensor-name>
   --- data
   --- dimensions
   --- name
   --- timestamps 

```

### TODO
- [x] Update CHANGELOG
- [x] Add example `xml` file to `app/xml`
- [x] Update CI for `yarp-telemetry`
- [ ] Update robology-superbuild's wearable dependency for `yarp-telemetry`

Closes #112

### CODE REVIEW

**Comment by prashanthr05** [comment]
I need to update the CI for `yarp-telemetry`.

**Comment by lrapetti** [comment]
Thanks a lot @prashanthr05, this is indeed a very useful tool that has always been missing.

Just a comment on your description

> It replaces,
> 
> > the `iwear_wrapper`'s logic of encapsulating the wearable sensors data into a WearableData msg and passing through a YARP port

the `iwear_remapper` is not replaced by itself, since it remains the device responsible for streaming iwear data on the network through yarp ports. It rather replaces the pipeline where the [`yarpdatadumper`](http://www.yarp.it/git-master/yarpdatadumper.html) was used to save those data.

**Comment by prashanthr05** [comment]
> Thanks a lot @prashanthr05, this is indeed a very useful tool that has always been missing.
> 
> Just a comment on your description
> 
> > It replaces,
> > > the `iwear_wrapper`'s logic of encapsulating the wearable sensors data into a WearableData msg and passing through a YARP port
> 
> the `iwear_remapper` is not replaced by itself, since it remains the device responsible for streaming iwear data on the network through yarp ports. It rather replaces the pipeline where the [`yarpdatadumper`](http://www.yarp.it/git-master/yarpdatadumper.html) was used to save those data.

True, I meant those lines at an implementation level and not at an architecture level. I will update the comment for clarity.

**Review by lrapetti — COMMENTED** [review]
Together with the sensor data, we can dump also the status of each sensor that is exposed with [`getSensorStatus`](https://github.com/robotology/wearables/blob/master/interfaces/IWear/include/Wearable/IWear/Sensors/ISensor.h#L102) method.

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (lrapetti) [reply wrappers/IWearLogger/src/IWearLogger.cpp:193]: Unfortunately, we are often using an `IWearRemapper` device in between the producer device, and the consumer device (in this case would be the `iwear_logger`), which merge data from multiple sources.

In this case, however, the state returned by `getStatus()` is the "worst status" that is found among all the associated sensors interface (see https://github.com/robotology/wearables/blob/master/devices/IWearRemapper/src/IWearRemapper.cpp#L802-L845). As a consequence, the specific status of the sensors interface is hidden. 

I think we can eventually discuss the current choice in the `IWearRemapper`, and change it.

**Comment by prashanthr05** [comment]
> > > Together with the sensor data, we can dump also the status of each sensor that is exposed with [`getSensorStatus`](https://github.com/robotology/wearables/blob/master/interfaces/IWear/include/Wearable/IWear/Sensors/ISensor.h#L102) method.
> > 
> > 
> > Do you think it would be a good idea to prefix the actual measurements with the sensor status?
> > Otherwise, if we would like it as a separate vector, we might have to propose a feature request in `yarp-telemetry`.
> 
> Maybe we can add it as a prefix, and meanwhile propose to add it in `yarp-telemetry` if it can be useful also for other applications.

Done in https://github.com/robotology/wearables/pull/113/commits/1b87ac37dda399cd15c43b55375b816eebfcaf29

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (prashanthr05) [reply wrappers/IWearLogger/src/IWearLogger.cpp:200]: These lines were a copy-paste from [iWearWrapper](https://github.com/robotology/wearables/blob/c6a8478343881468b91112244bd25c6a803d456e/wrappers/IWear/src/IWearWrapper.cpp#L151-L155). Do you think this might not be needed here?

**Comment by traversaro** [comment]
I don't know what's going on with the Windows CI. For me we can disable it and substitute it in a following PR with a conda-based CI that gets YARP in binary form.

**Comment by prashanthr05** [comment]
Apart from that, looks like CI Ubuntu checks on `iwear_logger` addition seems to be working now. We can wait for macOS for confirmation.

**Comment by prashanthr05** [comment]
> @prashanthr05 can you add the dependency of wearables on yarp-telemetry on the supernbuild? Thanks!

@traversaro , for this adding the line

``` cmake
find_or_build_package(YARP COMPONENTS telemetry QUIET)
```

to https://github.com/robotology/robotology-superbuild/blob/master/cmake/Buildwearables.cmake must suffice right?

**Comment by Nicogene** [comment]
@prashanthr05 the `find_or_build_package` should be like this

`find_or_build_package(YARP_telemetry)`

(See https://github.com/robotology/robotology-superbuild/blob/0a336d66d2d438243bdc12298767350460363cde/cmake/RobotologySuperbuildLogic.cmake#L51)

I am not sure if `COMPONENTS` works with that macro

**Comment by prashanthr05** [comment]
> @prashanthr05 the `find_or_build_package` should be like this
> 
> `find_or_build_package(YARP_telemetry)`
> 
> (See https://github.com/robotology/robotology-superbuild/blob/0a336d66d2d438243bdc12298767350460363cde/cmake/RobotologySuperbuildLogic.cmake#L51)
> 
> I am not sure if `COMPONENTS` works with that macro

Thanks for the heads up!! @Nicogene

File: wrappers/IWearLogger/src/IWearLogger.cpp
Comment (lrapetti) [reply wrappers/IWearLogger/src/IWearLogger.cpp:200]: > These lines were a copy-paste from [iWearWrapper](https://github.com/robotology/wearables/blob/c6a8478343881468b91112244bd25c6a803d456e/wrappers/IWear/src/IWearWrapper.cpp#L151-L155). Do you think this might not be needed here?

No, I think is fine as it is. I would rather change some logic on the Remapper side

**Review by lrapetti — APPROVED** [review]
I think we can proceed merging, and open a separate issue to move to conda-based Windows CI as @traversaro was suggesting

**Comment by lrapetti** [comment]
> I think we can proceed merging, and open a separate issue to move to conda-based Windows CI as @traversaro was suggesting

@prashanthr05 let me know if you want to clean the commit history or we can squash them and merge

**Comment by prashanthr05** [comment]
> > I think we can proceed merging, and open a separate issue to move to conda-based Windows CI as @traversaro was suggesting
> 
> @prashanthr05 let me know if you want to clean the commit history or we can squash them and merge

I have reorganized the commits.

**Comment by lrapetti** [comment]
Thanks @prashanthr05!

**Comment by diegoferigo** [comment]
Sorry for coming late to this PR, it looks good @prashanthr05! Though, it seems that your IDE didn't use clang-format. The next time someone is going to edit the file, a possibly big diff will get generated. If possible, I'd fix the style in a new PR. I'm not sure if recently the style has been enforced on all files @lrapetti @Yeshasvitvs.

**Comment by lrapetti** [comment]
> If possible, I'd fix the style in a new PR. I'm not sure if recently the style has been enforced on all files @lrapetti @Yeshasvitvs.

On my side, not really. Do you think it would make sense to open a PR to fix the style in the whole project?

**Comment by diegoferigo** [comment]
Until I was the maintainer of this project, I was enforcing the style strictly. With low priority, it could be done. I always though to start using github actions for that, but I still haven't found enough time to check what the community offers.

**Comment by diegoferigo** [comment]
Mmh ok, qtcreator automatically detects the `.clang-format` in the root of the repo and uses it for all the files. No idea about KDevelop.

I checked again the sources, and for sure the style is not uniform. Just check the if brackets, sometimes they are inline and some other time not. Be aware, when you start using clang-format there's no way back :laughing: I'm a bit nazi on it, for reasons :)

**Pull request merged**
